### PR TITLE
feat: ability to remove admins ❌

### LINKED ISSUE #250: Remove admin access in the Admin Dashboard 👥
### Description

See #247 for project overview. Note that this is dependent on #249 being implemented first.

### Acceptance Criteria

This issue should:
- Add a dropdown to the table in the `/admins` page, which has the option to remove access from the Admin Dashboard.
- Clicking the dropdown option should open a confirmation modal (ie: `/admins/remove`). In that modal, clicking a "Remove" button should officially remove access from the Admin Dashboard and they should no longer appear in the Admins table. Note that it shouldn't delete the record entirely, but it should set the `deletedAt` timestamp, effectively archiving the admin).

### Additional Context

  - You can add the backend functionality (ie: `removeAdmin`) in the `core` package. There is already an [admin](https://github.com/colorstackorg/oyster/tree/main/packages/core/src/modules/admin) module in the `core` package, you'll just need to add the function to the `use-cases` folder in that module.

### PR DESCRIPTION
**Pull request by wflore19**
## Description ✏️

Closes #250 

- Added `AdminDropdown` with the option to remove an admin to the `AdminsTable` component.
- Added `removeAdmin()` function to update an admins `deletedAt` column.
- Added modal to confirm the removal of an admin.

## Type of Change 🐞

- [x] Feature - A non-breaking change which adds functionality.
- [ ] Fix - A non-breaking change which fixes an issue.
- [ ] Refactor - A change that neither fixes a bug nor adds a feature.
- [ ] Documentation - A change only to in-code or markdown documentation.
- [ ] Tests - A change that adds missing unit/integration tests.
- [ ] Chore - A change that is likely none of the above.

## Checklist ✅

- [x] I have done a self-review of my code.
- [x] I have manually tested my code (if applicable).
- [ ] I have added/updated any relevant documentation (if applicable).

### CODE REVIEW

**Comment by ramiAbdou** [comment]
@wflore19 I know that the issue was stale, but please still comment on any issues that you'd like to take so other people know who's assigned to what 🙏

**Comment by wflore19** [comment]
@ramiAbdou heard, i think there's a potential issue with this pr, when removed from admins list i'm still logged in

**Comment by wflore19** [comment]
@tomas-salgado sounds good, I might need help with this one, how does authentication work for admins? Auth needs to be revoked when an admin is removed. Also can any admin remove another admin? What is the difference between temp admins vs FT admin?

**Comment by ramiAbdou** [comment]
> @tomas-salgado sounds good, I might need help with this one, how does authentication work for admins? Auth needs to be revoked when an admin is removed. Also can any admin remove another admin? What is the difference between temp admins vs FT admin?

Great question - we actually need to slightly rework our authentication/authorization mechanism in the Admin Dashboard. Will push an update and follow up a bit later today!

**Comment by ramiAbdou** [comment]
Hey @wflore19 - following up here, I was able to merge #426 yesterday which revamped some of the authentication logic in the Admin Dashboard. Here's what you need to know:
- Instead of just checking for the presence of a `user_id` in the session, now we actually query for the admin in the database. This ensures that if someone was logged into the Admin Dashboard, gets deleted and then goes back to the Admin Dashboard, they will be logged out despite previously being logged in. See [`getAuthenticationStatus`](https://github.com/colorstackorg/oyster/blob/main/apps/admin-dashboard/app/shared/session.server.ts#L91).
- I introduced a `role` to the `admins` table, which can be `'ambassador' | 'admin' | 'owner'`.
  - Owners can do anything.
  - Admins can add/remove other admins/ambassadors.
  - Ambassadors can't do anything.
- We can protect a route using `ensureUserAuthenticated` and the new [`minimumRole`](https://github.com/colorstackorg/oyster/blob/main/apps/admin-dashboard/app/shared/session.server.ts#L44) option.
- We also now have access to a hook called [`useAdminRole`](https://github.com/colorstackorg/oyster/blob/main/apps/admin-dashboard/app/shared/hooks.ts#L15), which allows you to use the role in front-end logic (ie: whether or not to display the "delete" button).

Also, just noting that I consolidated some of the admins module (we're going to less splitting up of files going forward and more colocating of related logic), so [`admin.core.ts`](https://github.com/colorstackorg/oyster/blob/main/packages/core/src/modules/admin/admin.core.ts) is where most business logic should live. Feel free to model your work after the `addAdmin` function!

Let me know if you have any questions at all!

File: packages/core/src/modules/admin/admin.core.ts [hunk 1]
```diff
@@ -156,6 +156,25 @@ export async function addAdmin({
...
+/**
+ * 'Removes' ColorStack admin. This will revoke the user access to the Admin
+ *  Dashboard. Note that it shouldn't delete the record entirely, but it should
+ *  set the deletedAt timestamp, effectively archiving the admin).
+ */
+export async function removeAdmin({ id }: { id: string }) {
+  const admin = await db
+    .updateTable('admins')
+    .set({
+      deletedAt: new Date(), // not sure of the format
```
Comment (tomas-salgado) [inline packages/core/src/modules/admin/admin.core.ts:168]: Yep, this looks like the correct format

**Review by tomas-salgado — COMMENTED** [review]
This looks great! Tested out the functionality and it all works correctly. 

The only thing I wonder is if we should implement the functionality so that you can't remove yourself as an admin 😆. I don't think there would ever be a reason to do this, and it would most likely just be a mistake.

**Comment by wflore19** [comment]
@tomas-salgado sounds good i'll make the change

**Comment by wflore19** [comment]
### Notes

In order to conditionally render the `delete` action i can either

- pass the `userId` prop from `AdminTable` -> `Table` -> `TableBody`
- refactor the `table.dropdown` component to pass an `id` prop, to conditionally render some options

**Comment by wflore19** [comment]
## Update

I see that all the data (list of admins) is being passed down to `<TableBody />` and renders the rows of data. So I would need to prop drill the current `adminsId` all the way down in order to conditionally render the `<Dropdown />`. This might be an issue as I would need to directly change `<TableBody />` specifically for the `<AdminTable />` implementation, which might not work for other implementations of `<TableBody />`.

**Comment by ramiAbdou** [comment]
@wflore19 the `Table.Dropdown` component actually has access to all the data for the row it's being rendered in -- notice how it's being rendered [here](https://github.com/colorstackorg/oyster/blob/d97cc3661cf29768d7d77727becdf3d8856d53ff/packages/ui/src/components/table.tsx#L174) (ie: `<Dropdown {...row} />`).

So in this case, if we wanted to omit the dropdown for certain rows, we can just `return null` early before rendering the actual dropdown (ie: `Dropdown.Container`)! Let me know if that makes sense.

**Comment by wflore19** [comment]
That makes sense, if the current users id EQUALS the id given from row data, return null. How should I handle passing the current users 'id' to the '<Dropdown \>', w/ useContext or prop drilling?

**Comment by wflore19** [comment]
Ok I implemented w/ `useContext()`. `<Dropdown />` conditionally renders if the `userId` is EQUAL to `id` 

admintable

**Review by tomas-salgado — APPROVED** [review]
LGTM! Great work @wflore19

**Review by ramiAbdou — APPROVED** [review]
Hey @wflore19 -- just tested this out and it looks really great! 🔥

Really appreciate you asking questions throughout this PR, it inspired an important change around the way that we handle auth, which is really great for us. Thanks for sticking through this PR even despite some of the uncertainty!

I made a few updates [here](https://github.com/colorstackorg/oyster/pull/408/commits/838191dfdc27e02cd721440a680833ad5c98dfcf) that revolve around whether or not the admin has permission to remove another admin. Check it out and let me know if you want to talk through some of those changes.

Another Wilfredo PR in the books! 🚀

**Pull request merged**